api/client/proto/authservice.proto

```diff
@@ -150,6 +150,40 @@ message Certs {
     bytes SSH = 1 [ (gogoproto.jsontag) = "ssh,omitempty" ];
     // TLS X509 cert (PEM-encoded).
     bytes TLS = 2 [ (gogoproto.jsontag) = "tls,omitempty" ];
+    // TLSCACerts is a list of TLS certificate authorities.
+    repeated bytes TLSCACerts = 3 [ (gogoproto.jsontag) = "tls_ca_certs,omitempty" ];
+    // SSHCACerts is a list of SSH certificate authorities.
+    repeated bytes SSHCACerts = 4 [ (gogoproto.jsontag) = "ssh_ca_certs,omitempty" ];
+}
+
+// GenerateServerKeysRequest specifies certificate-generation parameters
+// for a server.
+message GenerateServerKeysRequest {
+    // HostID is a unique ID of the host.
+    string HostID = 1 [ (gogoproto.jsontag) = "host_id" ];
+    // NodeName is a user-friendly host name.
+    string NodeName = 2 [ (gogoproto.jsontag) = "node_name" ];
+    // Roles is a list of roles assigned to the host.
+    repeated string Roles = 3 [ (gogoproto.jsontag) = "roles" ];
+    // AdditionalPrincipals is a list of additional principals
+    // to include in OpenSSH and X509 certificates
+    repeated string AdditionalPrincipals = 4
+        [ (gogoproto.jsontag) = "additional_principals,omitempty" ];
+    // DNSNames is a list of DNS names to include in x509 certificates.
+    repeated string DNSNames = 5 [ (gogoproto.jsontag) = "dns_names,omitempty" ];
+    // PublicTLSKey is a PEM encoded public key used for TLS setup,
+    bytes PublicTLSKey = 6 [ (gogoproto.jsontag) = "public_tls_key" ];
+    // PublicSSHKey is a SSH encoded public key.
+    bytes PublicSSHKey = 7 [ (gogoproto.jsontag) = "public_ssh_key" ];
```

Comment:
Are the Public TLS / Public SSH the keys signed? Are they optional? Do they have to match? If they have to match, can't we just generate the SSH key from the TLS one?

Please document the answers.

Reply:
They are not yet signed. This message is essentially a CSR - you send the auth server some public keys and it sends back signed certificates.

They are both required. We used to support omitting them (both) and allowing the auth server to generate the private key and corresponding public keys, but that behavior never actually worked outside of test code due to a bug and was recently removed in #8135. 

In all cases I'm aware of, clients generate these from the same private key, but the auth server doesn't know or cares whether or not that is true  - it's just going to return signed certificates to the client.

I will take a stab at adding this detail to the proto message.

Reply:
Right, thanks.